graphql-metrics 3.0 rewrite w/ AST analyzer, instrumentation and tracing

chrisbutcher opened the pull request:
~This PR is still technically WIP because of the arguments-related issues described below. But otherwise the query and fields metrics related code and tests in test/unit/analyzer_test.rb are now ready for a preliminary review from @Shopify/api-patterns-team~

~I've investigated a few workarounds to problems ¹ & ² (described below) but those partial solutions (not pushed here) started getting very complex, so I've paused until I understand https://github.com/rmosolgo/graphql-ruby/issues/2573 more.~

# TODOs
* [x] Update README.md
* [x] Switch instrumentation and tracer to instances
* [x] Figure out if the tracer's yielding method structure is clear, or if we want to reference / pass blocks explicitly. edit: I think this is OK now.

# Nonblocking TODOs
* [ ] Figure out GraphQLMetrics.use options, options passed into instrumentation, analyzer, tracer - Can maybe defer to gem v3.0.1
  * [ ] Blocker to using `.use` written up, here: https://github.com/rmosolgo/graphql-ruby/pull/2614

# Update Nov 12

This PR is now ready for review. I've unblocked myself with regard to arguments logging, by reverting the use of `on_leave_argument`, and simply recursing arguments manually via `on_leave_field`.

# Progress

* [x] Query metrics
  * [x] `operation_type` e.g. "mutation"
  * [x] `operation_name` e.g. "PostCreate"
  * [x] `start_time` (unix timestamp), per Apollo spec
  * [x] `duration` (in seconds, captured using monotonic clock)
* [x] Field metrics
  * [x] `field_name` e.g. "id"
  * [x] `return_type_name` e.g. "ID"
  * [x] `parent_type_name` e.g. "Post"
  * [x] `deprecated` e.g. false
  * [x] `path` e.g. ["postCreate", "post", "id"]
  * [x] `resolver_timings`
    * [x] `start_time` (unix timestamp)
    * [x] `duration` (float, seconds)
  * [x] `lazy_resolver_timings`, same structure as resolver_timings
* [x] Arguments ~¹~
  * [x] `argument_name` e.g. "id"
  * [x] `argument_type_name` e.g. "ID"
  * [x] `parent_field_name` e.g. "post"
  * [x] `parent_field_type_name` e.g. "QueryRoot"
  * [x] `default_used` ~²~ e.g. false
  * [x] `value_is_null` ~²~ e.g. false
  * [x] `value` ~²~ e.g. GraphQL::Query::Arguments::ArgumentValue (we use this to get values `value.definition.metadata` in Shopify).

~¹ = Arguments are partially captured in this branch, including any top-level args specified inline or with variables. Input object fields usage is not captured when input objects are specified by variables.~
~² = These argument metrics are NOT currently captured for nested input objects, even if specified inline (and not with variables).~

~These blockers are described in https://github.com/rmosolgo/graphql-ruby/issues/2573~

swalkinshaw commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,179 @@
+module GraphQLMetrics
+  class Analyzer < GraphQL::Analysis::AST::Analyzer
```
I know this is WIP but a couple ideas:

* having one file is nice, but it might make things more confusing. Especially for naming?
* `Analyzer` is also a tracer right now which is a symptom of the above
* whatever this is called, it would be great to offer a single `use` method so it can act as a plugin:

```ruby
class MySchema < GraphQL::Schema
  use GraphQL::Metrics
end
```

The `use` method can handle all of this:

```ruby
instrument :query, SimpleAnalyzer
query_analyzer SimpleAnalyzer
tracer SimpleAnalyzer
```

If someone wants to opt-out of one, then they'd just do that manually by picking whatever parts they want.

chrisbutcher replied:
💯 !

chrisbutcher replied:
> having one file is nice, but it might make things more confusing. Especially for naming?

I was also thinking of the option of keeping it as one file, and just having private modules (for tracing, instrumentation methods) that get included into the analyzer class.

swalkinshaw replied:
Whatever makes it easier for development purposes 👍 The main thing is providing a simpler way for consumers to implement this. So at least they'd have only a single "entry point" by default.

swalkinshaw replied:
🤔 I might be missing something (and/or jetlagged) but the gem automatically instantiates the analyzers: https://github.com/swalkinshaw/graphql-ruby/blob/1149e07bd8b5f64f4b1bd7265d95fdeb77d6baa5/lib/graphql/analysis/ast.rb#L60

So your initializer will always be run without having to do anything manually?

chrisbutcher replied:
graphql-ruby does initialize analyzers by default when they are provided like so:

```ruby
  class Schema < GraphQL::Schema
    # ...
    query_analyzer MyAnalyzer
  end
```

But it doesn't do so when they are provided via `schema_defn.query_analyzer(Analyzer)`, as seen in the `self.use` example above.

swalkinshaw replied:
I don't think thats true since both methods just add an analyzer to the `query_analyzers` array? Either way they get instantiated later in the file I linked.

chrisbutcher replied:
I've push up this commit on a throwaway branch to illustrate the issue:

https://github.com/Shopify/graphql-metrics/commit/ff6f93bb74337d6baec9b75fc850015d07925b18

chrisbutcher replied:
Thanks @swalkinshaw for spotting the issue:

```diff
-  def self.use(schema_defn)
+  def self.use(defn)
+    schema_defn = defn.target
     schema_defn.instrument(:query, Instrumentation)
-    schema_defn.query_analyzer(Analyzer)
-    schema_defn.tracer(Tracer)
+    schema_defn.query_analyzers << Analyzer
+    schema_defn.tracers << Tracer
   end
```

swalkinshaw replied:
This still doesn't work unfortunately. We dug into it and it's definitely an upstream bug due how the gem handles the legacy layer under the class-based one.

chrisbutcher replied:
Yeah I'll repro it upstream and hope to fix it / see it fixes by the time we want to cut the 3.0 gem publicly.

eapache commented:
hmm, not ideal for that ping to show up on a public repo @chrisbutcher

eapache commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+      def trace_field(_key, data, block, context_key)
+        path_excluding_numeric_indicies = data[:path].select { |p| p.is_a?(String) }
+
+        start_time = current_time
```
as per the global one, do we need both times here? clock lookups are expensive, and doing 2 per field (vs 1 per field) is probably non-trivial though I have no benchmarks to back that up.

chrisbutcher replied:
I believe we need both wall and monotonic clock times: https://github.com/Shopify/graphql-metrics/pull/10#discussion_r341349517

eapache replied:
I can see the argument for once per query, but I don't believe we need wall clock times per field, since we can effectively calculate this later by comparing the monotonic time with the start-of-query monotonic time. And in fact that's what the apollo spec asks for anyway, is the `startOffset` monotonic difference.

chrisbutcher replied:
Oh excellent point. Will fix.

eapache commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+
+      def after_query(query)
+        return if query.context[:skip_graphql_metrics_analysis]
+        return unless query.valid?
```
this is already checked inside `analyze?`, won't the gem only call before/after if `analyze` returns true?

chrisbutcher replied:
I can scrap this check here, and only do it in `initialize` or the idempotent method I speculated about above.

eapache commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+      query_extracted(@static_query_metrics.merge(runtime_metrics))
+    end
+
+    # TODO: Apollo Tracing spec https://docs.google.com/document/d/1B0dR09CcN_M4yqezkJ7VFPI-mKgQIVCt9PiUcIMEVNI/edit
```
eh?

chrisbutcher replied:
I'll delete this comment. It was just a reminder to ensure the data we capture is compliant with the Apollo spec out of the box.

eapache commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+
+        data[:query].context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          ns[context_key][path_excluding_numeric_indicies] ||= []
+          ns[context_key][path_excluding_numeric_indicies] << { start_time: start_time, duration: duration }
```
does this mean we're collecting the same field referenced through different paths under different events? that seems like it would unnecessarily inflate our event count and make aggregation harder?

chrisbutcher replied:
It just means that a given field, resolved n times within a list, connection etc. would be attributed n (multiplied for nested lists/connections) timings.

So:

```graphql
products(first: 5) {
  edges {
    node {
      title
      variants(first: 2) {
        title
      }
    }
  }
}
```

We extract fields only once, but in this example we may have 5 timings attributed to `shop.products.edges.node.title`, and 10 timings attributed to `shop.products.edges.node.variants.edges.node.title`.

eapache commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+      }
+
+      if self.class.runtime_metrics_enabled?(visitor.query.context)
+        callback = -> (metrics) {
```
perhaps a micro-optimization, but right now we're recreating this callback on every field, where I think it's going to be identical for every one? the only out-of-closure reference is to `visitor` which I *think* is constant across the whole query?

chrisbutcher replied:
Good point! I don't think this was always the case before recent changes. I'll revisit this decision to store a callback per field visited.

eapache commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+          end
+        }
+
+        @field_timing_callbacks << FieldTimingCallback.new(static_metrics: static_metrics, callback: callback)
```
flatten `static_metrics` into `FieldTimingCallback` to save a hash allocation per field?

eapache commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+      end
+    end
+
+    def run_field_timing_callbacks
```
gotta admit I don't fully grok the whole callback setup you've got going on here

chrisbutcher replied:
tldr, when used as instrumentation, an analyzer and tracing, the order of execution is:

* before_query (context setup)
* initialize (bit more context setup, instance vars setup)
  ^ I'm reconsidering where and how the setup logic is all done
* result (static analyzer results ready, including the opportunity to call extract_query for static query metrics if the analyzer is not being used as instrumentation)
* after_query (call query and field callbacks, now that we have all static and runtime metrics gathered)
* extract_query
* run_field_timing_callbacks
* calling field timing callback (n times)

When used as a simple analyzer, which doesn't gather or emit any runtime metrics (timings, arg values):
* initialize
* result
* extract_query

eapache commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+        callback = -> (metrics) {
+          visitor.query.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+            resolver_timings = ns[GRAPHQL_TRACING_FIELD_KEY][metrics[:path]]
+            lazy_field_timing = ns[GRAPHQL_TRACING_LAZY_FIELD_KEY][metrics[:path]]
```
🤔 I don't see anything that populates the lazy value...

chrisbutcher replied:
This is populated via `trace` and `trace_field`, https://github.com/Shopify/graphql-metrics/blob/wip-ast-interpreter-rewrite/lib/graphql_metrics/analyzer.rb#L79.

I think your comment about the double use of the constants will make this more clear.

chrisbutcher commented on Gemfile:
```diff
@@ -5,6 +5,8 @@ git_source(:github) {|repo_name| "https://github.com/#{repo_name}" }
 # Specify your gem's dependencies in graphql_metrics.gemspec
 gemspec
 
+gem 'graphql', '1.9.14', path: '../../rmosolgo/graphql-ruby'
```
Sorry I didn't mean to push this change :) It's just for local debugging in the graphql ruby gem.

chrisbutcher commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+    FieldTimingCallback = Struct.new(:static_metrics, :callback, keyword_init: true)
+
+    CONTEXT_NAMESPACE = :graphql_metrics_analysis
+    RUNTIME_METRICS_ENABLED = :runtime_metrics_enabled
```
It's a key used in the `graphql_metrics_analysis` context namespace in order to make code that otherwise expects runtime values still work (by checking `runtime_metrics_enabled?`) when the analyzer is used without instrumentation/tracing (in other words, when someone just wants static metrics).

chrisbutcher commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+    FIELD_TIMING_CALLBACKS_KEY = :field_timing_callbacks
+    ANALYZER_INSTANCE_KEY = :analyzer_instance
+
+    # NOTE: These constants are used both to match tracing keys from graphql-ruby as well as to store in-line and lazy
```
Good question. It's two additional constants, but probably more clear, and if the underlying tracing key values in the gem change, we don't have to fork the constants at that point.

I'll split them up.

chrisbutcher commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+          ns[RUNTIME_METRICS_ENABLED] = true
+          ns[FIELD_TIMING_CALLBACKS_KEY] = []
+          ns[QUERY_START_TIME_KEY] = current_time
+          ns[QUERY_START_TIME_MONOTONIC_KEY] = current_time_monotonic
```
We want wall clock time for query / field execution start times, since it's UNIX time (seconds since the UNIX epoch, Jan 1 1970).

We want monotonic clock time for determining durations. Monotonic clocks (at least in Ruby) return the seconds since the machine booted. So I think we need both.

```bash
$ ruby -e "now = Process.clock_gettime(Process::CLOCK_MONOTONIC); puts now; puts \"#{now / 86400} days\""
102546.841438
1.1868847388657409 days
$ uptime
16:23  up 1 day,  4:29, 8 users, load averages: 2.18 1.92 1.72
```

Also see https://github.com/apollographql/apollo-tracing#collected-data

chrisbutcher commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+
+      def after_query(query)
+        return if query.context[:skip_graphql_metrics_analysis]
+        return unless query.valid?
```
It is indeed: https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/query/validation_pipeline.rb#L34-L37

chrisbutcher commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+          duration = current_time_monotonic - ns[QUERY_START_TIME_MONOTONIC_KEY]
+          runtime_metrics = { start_time: ns[QUERY_START_TIME_KEY], duration: duration }
+
+          analyzer = ns[ANALYZER_INSTANCE_KEY]
```
Hmm good question.

I went this route because I want to support this analyzer being used on its own, without instrumentation (which calls `before_query`). But now that I think of it, I could probably scrap the logic in `initialize` and do it idempotently either in `before_query` (when instrumentation is being used) or in the first analyzer hook method that notices that setup wasn't done (when not used as instrumentation).

chrisbutcher replied:
I can confirm the gem does create new analyzer instances each time query.result is called.

I thought through options to move setup logic currently in before_query and initialize to other places, but decided to largely leave it as is (haven't pushed local commits yet, will notify when I do):

`before_query` has all of the runtime metrics context key-value setup, since we only need this stuff when instrumentation is opted into.

`initialize` seems like the best place to obtain the analyzer instance to store for later use, as well as the instance vars that no instrumentation hooks need access to.

grcooper commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+module GraphQLMetrics
+  class Analyzer < GraphQL::Analysis::AST::Analyzer
```
It feels like there are two different operations happening inside of this class:
1. The analyzer is adding callbacks for metrics per field
2. There is a setup beforehand that creates the callbacks etc. Should these two concepts be split into different classes?

swalkinshaw commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,193 @@
+    end
+
+    def analyze?
+      query.valid? && query.context[:skip_graphql_metrics_analysis] != true
```
@eapache commented about this in another place, but the analyzer is only run if the query is valid anyway. Unless you want this if someone manually runs the analyzer? But that's an odd edge case to account for

chrisbutcher replied:
Yeah I initially decided to support the ad hoc use case, since it seems fairly useful to be able to run extract metrics out of band from production queries. This cached query validity checks, as well the calls to `timings_capture_enabled?` are there to support the analyzer being invoked directly.

chrisbutcher commented:
This is still WIP. I'm going to take another run at a workaround to get fully arguments (nested input objects, specified by variables etc.) logging working before I request another round of reviews.

chrisbutcher commented:
Updated and ready for another review.

grcooper commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,140 @@
+    end
+
+    def analyze?
+      query.valid? && query.context[GraphQLMetrics::SKIP_GRAPHQL_METRICS_ANALYSIS] != true
```
Can't this be simplified to `query.valid? && !query.context[GraphQLMetrics::SKIP_GRAPHQL_METRICS_ANALYSIS]`? Rather than adding the `!= true`?

chrisbutcher replied:
Yeah that's better actually, since we may as well embrace that context value being truthy, rather than `true`.

grcooper commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,140 @@
+        extract_arguments(value.value, field_defn)
+      when ::GraphQL::Schema::InputObject
+        extract_arguments(value.arguments.argument_values.values, field_defn)
+      end
```
Is there ever any default case? Should we error if it isn't one of these types?

chrisbutcher replied:
Good question! The only scenario that would enter the `else` block, if we had one would be when `value` is an argument scalar value, which we don't care about, since:

```ruby
when ::GraphQL::Query::Arguments::ArgumentValue
  extract_argument(value, field_defn) #  we log all that we need about scalar arguments here
  extract_arguments(value.value, field_defn) # we recurse here anyway, regardless of whether `value.value` is a nested array, hash, set of arguments, argument value or scalar. If it's a scalar, we no-op rather than handling it via `else`.
```

chrisbutcher commented on test/unit/graphql_metrics_integration_test.rb:
```diff
@@ -0,0 +1,997 @@
+require "test_helper"
+require "test_schema"
+
+class GraphQLMetricsIntegrationTest < ActiveSupport::TestCase
```
Commenting here, to ensure Github doesn't hide this file from reviewers by default

swalkinshaw replied:
it does anyway lol 😞  still need to manually expand it

chrisbutcher replied:
nooooooooooooooo

eapache commented on graphql_metrics.gemspec:
```diff
@@ -33,16 +33,19 @@ Gem::Specification.new do |spec|
   spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
   spec.require_paths = ["lib"]
 
+  spec.add_runtime_dependency "concurrent-ruby", "~> 1.1.0"
```
?

chrisbutcher replied:
This is explained here: https://github.com/Shopify/graphql-metrics/pull/10/files#diff-eb87a16ba8d7a5b04c0484ae1f802b83R5-R7

I suppose I could use `Thread.current[]=`? I figured the concurreny-ruby gem is preferable due to its maturity.

Is it silly to worry about the same tracer running concurrently?. I'm not sure if multiplexing in the gem can / will support doing so across multiple queries at once in the future. It seemed safer to do this, to keep values from the lexing, parsing phase before we have a context to store values in for later.

chrisbutcher replied:
Confirmed, this library ultimately uses `Thread.current[]=` (just like tracers for graphql-ruby for Skylight, NewRelic), but with better guarantees (and some slowdown, hopefully negligible) due to locking w/ a Mutex.

https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent/atomic/ruby_thread_local_var.rb

eapache commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,140 @@
+    def initialize(query_or_multiplex)
+      super
+
+      query_or_multiplex.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
```
this doesn't really need to be a `tap`?

eapache commented on lib/graphql_metrics/instrumentation.rb:
```diff
@@ -1,109 +1,40 @@
+
+      query.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+        query_duration = GraphQLMetrics.current_time_monotonic - ns[GraphQLMetrics::QUERY_START_TIME_MONOTONIC]
+        query_end_time = GraphQLMetrics.current_time
```
this isn't really a useful value since we already have the start time *and* the duration?

chrisbutcher commented on test/test_schema.rb:
```diff
@@ -6,22 +6,33 @@ def initialize(model)
   end
 
   def perform(ids)
+    sleep 1
```
I left them in since there are so few tests, everything still runs in seconds, and we can verify that >= 1 second was spent in batch loading. It's mainly like this for ease for understanding the tests.

chrisbutcher commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,140 @@
+
+    def extract_fields_with_runtime_metrics(context)
+      @static_field_metrics.each do |static_metrics|
+        context.namespace(CONTEXT_NAMESPACE).tap do |ns|
```
Yeah, it initially felt a bit cleaner to do .tap everywhere that i read/wrote to the context, but the additional nesting isn't worth the "code organization" that I was really getting from it.

Removing all `.tap`s.

chrisbutcher commented on lib/graphql_metrics/tracer.rb:
```diff
@@ -0,0 +1,128 @@
+      return setup_tracing_before_lexing(resolver_block) if key == GRAPHQL_GEM_LEXING_KEY
+      return capture_parsing_time(resolver_block) if key == GRAPHQL_GEM_PARSING_KEY
+
+      context = possible_context || data[:multiplex].queries.first.context
```
Yes, after parsing, we have a context during validation and all later phases.

swalkinshaw commented on lib/graphql_metrics/tracer.rb:
```diff
@@ -0,0 +1,125 @@
+      skip_tracing = possible_context&.fetch(GraphQLMetrics::SKIP_GRAPHQL_METRICS_ANALYSIS, false)
+      return yield if skip_tracing
+
+      return setup_tracing_before_lexing { yield } if key == GRAPHQL_GEM_LEXING_KEY
```
🤔 this works? I wasn't really clear but you can still do `setup_tracing_before_lexing(&resolver_block)`. Just need `&`

chrisbutcher replied:
Yeah, I'm just yielding all the way through now without any & referenced blocks. I think it still reads fairly nicely.

chrisbutcher commented:
Updated.

swalkinshaw commented on test/unit/graphql_metrics_integration_test.rb:
```diff
@@ -0,0 +1,995 @@
+    )
+  end
+
+  class Schema2 < GraphQL::Schema
```
What's the difference here? Maybe name them more semantically? I'd also probably just define this up top too

chrisbutcher replied:
I'll rename them. One schema gets both static and runtime metrics, and one (Schema2) gets only static metrics to demonstrate this is all possible if someone wants usage but not timings.

swalkinshaw commented on test/unit/graphql_metrics_integration_test.rb:
```diff
@@ -0,0 +1,995 @@
+    GRAPHQL
+  end
+
+  # TODO: Need nested inputs to test argument value extraction here, as well as in Shopify/shopify
```
still a todo?

chrisbutcher replied:
Nope I'll delete this comment.

swalkinshaw commented on lib/graphql_metrics/tracer.rb:
```diff
@@ -0,0 +1,125 @@
+      possible_context = data[:query]&.context
+
+      skip_tracing = possible_context&.fetch(GraphQLMetrics::SKIP_GRAPHQL_METRICS_ANALYSIS, false)
+      return yield if skip_tracing
```
After this early return I find the rest of the method a bit hard to follow. I wonder if it would read better as a big case statement the way the gem does? https://github.com/swalkinshaw/graphql-ruby/blob/1149e07bd8b5f64f4b1bd7265d95fdeb77d6baa5/lib/graphql/tracing/platform_tracing.rb#L21-L61

chrisbutcher replied:
Great idea!

chrisbutcher commented on Rakefile:
```diff
@@ -5,6 +5,12 @@ Rake::TestTask.new(:test) do |t|
+  t.warning = false
+
+  puts "Reminder: Remove `t.warning = false` in Rakefile once graphql-ruby fixes all instances of"\
+    "`warning: instance variable @<ivar> not initialized` and `mismatched indentations`"
```
I bumped rubies and started getting many hundreds of these warnings, coming from graphql-ruby. This is left as a TODO. It's a bit unfortunate, but not doing so makes test output unnecessarily long.

RobertWSaunders replied:
Can confirm I am having the same behaviour in the test helper gem I am working on. If you find a solution I'd love to know it. For now I'm doing the same thing as you here. 😃 👍

chrisbutcher commented:
> I admit I don't fully understand the tracer yet.

I cleaned up the tracer a bit more, hopefully it's easier to grok.

> Especially the "pre context" stuff so I may take a deeper look at some point but I don't think it should hold this up

The tldr here is simply that I want to capture timings before the query context is available to store things in (`data[:query]` is not available yet during lexing, parsing). I chose to put these values in an OpenStruct, in a class attribute. I wrapped that OpenStruct in a `Concurrent::ThreadLocalVar` to not have to worry about concurrently handled / analyzed queries stepping on each others' "pre-context" (i.e. pre query validation) values.

chrisbutcher commented:
@RobertWSaunders I updated the README :)

eapache commented on lib/graphql_metrics/analyzer.rb:
```diff
@@ -0,0 +1,138 @@
+      return if GraphQLMetrics.timings_capture_enabled?(@query.context)
+
+      # NOTE: If we're running as a static analyzer (i.e. not with instrumentation and tracing), we still need to
+      # flush static query metrics somewhere other than `after_query`.
```
why?

eapache commented on lib/graphql_metrics/instrumentation.rb:
```diff
@@ -1,109 +1,37 @@
-
-      [start_time, self.class.current_time]
+      ns = query.context.namespace(CONTEXT_NAMESPACE)
+      query_duration = GraphQLMetrics.current_time_monotonic - ns[GraphQLMetrics::QUERY_START_TIME_MONOTONIC]
```
I would expect to see `QUERY_START_TIME_MONOTONIC` set in `before_query` for symmetry, but it's handled in the tracer for some reason?

chrisbutcher replied:
From the top of `analyzer.rb`

```ruby
# * Tracer.setup_tracing_before_lexing
# * Tracer.capture_parsing_time
# * Instrumentation.before_query (context setup)
```

I do this in `setup_tracing_before_lexing` because `before_query` doesn't actually run first, and tracing is our first opportunity to truly start these clocks at the beginning of graphql-ruby's phases of execution.

eapache commented on lib/graphql_metrics/tracer.rb:
```diff
@@ -0,0 +1,127 @@
+      # chronological order.
+      case key
+      when GRAPHQL_GEM_LEXING_KEY
+        return setup_tracing_before_lexing { yield }
```
I wonder if it is more efficient to materialize the block with `&block` and just pass it along, or if it's more efficient to do this and instantiate a second block just to yield to the first.

chrisbutcher replied:
I had block references, passed them as variables and used `.call` in my earlier iteration. @swalkinshaw suggested this approach and hinted it may be faster.

Digging into it a bit: even after some recent optimizations to passing `&block` and calling `.call`, `yield`ing directly still appears faster: https://github.com/JuanitoFatas/fast-ruby#proc--block

> or if it's more efficient to do this and instantiate a second block just to yield to the first.

I'm not sure I understand this suggestion. Can you provide a snippet?

swalkinshaw replied:
Before you were doing `resolve_block` without `&`. My original suggestion was to pass them as a block reference `&resolve_block` and `yield` still works and there's no need to explicitly `call` them

chrisbutcher replied:
Ah OK. Somehow I'd missed that.

I'm open to suggestions here. I suppose just `yield`ing everywhere without any block references is a bit more cryptic to readers?

eapache commented on lib/graphql_metrics/tracer.rb:
```diff
@@ -0,0 +1,127 @@
+      when GRAPHQL_GEM_PARSING_KEY
+        return capture_parsing_time { yield }
+      when *GRAPHQL_GEM_VALIDATION_KEYS
+        context = possible_context || data[:multiplex].queries.first.context
```
why would we assume that it's always the first query of the multiplex here?

chrisbutcher replied:
Oops. Will fix!

chrisbutcher replied:
We're also not explicitly handling multiplexes in analyzer.rb

I'll write a test to ensure that everything works as expected if we receive a multiplex instead of a single query. That said, if supporting multiplexes becomes complex, we could opt to support them in gem version 3.1.0.

chrisbutcher replied:
Wrote a small summary of why we're OK to ignore multiplexes from the perspective of this gem: https://github.com/Shopify/graphql-metrics/pull/10#issuecomment-555186588

chrisbutcher commented on README.md:
```diff
@@ -33,132 +33,170 @@ Or install it yourself as:
+      # Or store them on the query context:
+      store_metrics(:queries, metrics.merge(custom_metrics_from_context))
+
+      # For use after calling query.result, in this case by digging them out, for example in a Rails controller:
```
I originally had a toy Controller/def <action> in there, I'll re-add.

chrisbutcher commented on README.md:
```diff
@@ -33,132 +33,170 @@ Or install it yourself as:
-      used_in_operation: metrics[:used_in_operation],     # true
-    })
-  end
+(TODO, what options should contain).
```
Yeah I'll leave this as-is for now until I convert the instrumentation / tracer to instances.

chrisbutcher commented:
OK so I figured out a way forward re: multiplexing vs. handling things as if there's just one query (as this PR currently does).

tldr: We can safely ignore handling anything in this gem to do with multiplexing, as long as we're OK with ignoring time spent in any future tracing code that does any heavy lifting during the `analyze_multiplex` phase.

I revisited https://graphql-ruby.org/queries/multiplex.html to be sure of a few things. All queries executed by graphql-ruby are run through multiplexes. Usually there's just one i.e. `[query]`.

When more than one is specified and run as a multiplex, i.e. `[query, query, ... query]` a few things happen. First, note that graphql-ruby will execute all queries in a multiplex in serial. You can define instrumentation, analyzers and tracing that will handle multiplexes as an array, or simply handle events (before_query, after_query, on_leave_field, trace) assuming there's only one query, and without any awareness of what the "current" query is.

The way graphql-metrics 3.0 has been written so far, all state is setup/torn down appropriate between queries, so we're safe to ignore multiplex events, and treat all queries in a multiplex the same way as we do one-off queries from multiple HTTP calls. I was able to drop the one place I worked myself into a corner i.e. `data[:multiplex].queries.first.context` by simply ignoring `analyze_multiplex` events in the tracer.

```diff
- GRAPHQL_GEM_VALIDATION_KEYS = ['validate', 'analyze_query', 'analyze_multiplex']
+ GRAPHQL_GEM_VALIDATION_KEYS = ['validate', 'analyze_query']
```

The only downside to ignoring `analyze_multiplex` is that if we / anyone do a heavy amount of processing when `analyze_multiplex` tracing events are detected, we won't count it as part of the time spent validating the query. I personally think this is worth returning to at a later date.

chrisbutcher commented:
I added test coverage for when queries are executed as a multiplex, to ensure we still do everything correctly 👍. Tests are all passing.

Updated the TODOs in the PR description with the two things I think still need to be done on this PR.

chrisbutcher commented:
Thanks @swalkinshaw for the suggestion to introduce an abstraction, TimedResult, around all of the places we wrap yields in start times, offsets, durations.

chrisbutcher commented on README.md:
```diff
@@ -33,132 +33,174 @@ Or install it yourself as:
+
+      # For use after calling query.result, in this case by digging them out, for example in a Rails controller:
+      # class GraphQLController < ...
+      #   def graph_ql
```
I cheated here a bit. For some reason `def graphql` blows up my VsCode syntax highlighting in this file. Meant to reset this, Thanks!

chrisbutcher commented on README.md:
```diff
@@ -33,132 +33,174 @@ Or install it yourself as:
+    #     start_time_offset: 0.011901999998372048,
+    #     duration: 5.999987479299307e-06}
+    #   ],
+    #   lazy_resolver_timings: nil,
```
The latter, but I could show an example of that, too

chrisbutcher commented:
👋 @Shopify/api-patterns-team This PR is now ready for review again.

Since the last round of reviews, only a few few things have changed since last review.

One major one is that we're skipping logging for introspection fields. Secondly, GraphQLMetrics is also now properly namespaced -> GraphQL::Mertrics::*

One non-blocking TODO which is to make use of `GraphQL::Metrics.use`. which should work as of graphql-ruby 1.10.0, but that's not yet used in Shopify/shopify, and it appears graphql-metrics needs to be updated to work with that version, too (tldr: `schema_defn.target` has been removed upstream, so this breaks currently: https://github.com/Shopify/graphql-batch/blob/75bfef9b98faa15b59ee5df7f6ba1253083d7ee2/lib/graphql/batch.rb#L19)

eapache commented:
> tldr: schema_defn.target has been removed upstream, so this breaks currently

I think that's because 1.10 *only* has the class-based version, so you simply don't need to dig down into it anymore, you should get the class you need directly in the argument.

eapache commented:
There's a *lot* of code here. At what point is it better to start splitting this up and shipping it to master in pieces? We don't have to release the 3.0 gem until all the pieces have merged.

eapache commented on README.md:
```diff
@@ -33,132 +33,177 @@ Or install it yourself as:
+    # Tracing spec referred to above.
+    #
+    # `resolver_timings` is populated any time a field is resolved (which may be many times, if the field is nested
+    # within a list field e.g. a Relay connection field). There should always be at least one of these per field in
```
Why is there always at least one? `{ foo { id } }` but `foo` is null then `id` will be extracted with no resolutions.

chrisbutcher replied:
Good catch. That's 100% true. Added a test to confirm.

eapache commented on README.md:
```diff
@@ -33,132 +33,177 @@ Or install it yourself as:
+    #   path: ["post", "id"],
+    #   resolver_timings: [
+    #     start_time_offset: 0.011901999998372048,
+    #     duration: 5.999987479299307e-06}
```
this `}` looks malformed? and also in the lazy array

eapache commented on README.md:
```diff
@@ -33,132 +33,177 @@ Or install it yourself as:
-      used_in_operation: metrics[:used_in_operation],     # true
-    })
-  end
+(TODO, what options should contain).
```
Still TODO

chrisbutcher replied:
Yeah, I'll drop this any all mentions of `use GraphQL::Metrics` in this PR until it's unblocked in graphql-ruby. I'll leave a note for us in Shopify/shopify about it though.

eapache commented on lib/graphql/metrics.rb:
```diff
@@ -0,0 +1,70 @@
+        @start_time = GraphQL::Metrics.current_time_monotonic
+        @result = yield
+        @duration = GraphQL::Metrics.current_time_monotonic - @start_time
+        @time_since_offset = @start_time - @offset_time if @offset_time
```
Not entirely clear to me just from reading what the time_since_offset is from, or how/when offset_time would differ from start_time

chrisbutcher replied:
https://github.com/apollographql/apollo-tracing#response-format

This is intended to produce field resolver start offsets. In our case, field resolver start offsets are relative to the parent operation's start time. I'll reconsider these names, and add comments until it's clear.

|-------|

chrisbutcher replied:
I added a more thorough comment here:

https://github.com/Shopify/graphql-metrics/pull/10/files#diff-826511089b3e970c9ab2d7feebbf99fdR52-R69

eapache commented on lib/graphql/metrics/instrumentation.rb:
```diff
@@ -0,0 +1,48 @@
+        return if query.context[GraphQL::Metrics::SKIP_GRAPHQL_METRICS_ANALYSIS]
+
+        ns = query.context.namespace(CONTEXT_NAMESPACE)
+        query_duration = GraphQL::Metrics.current_time_monotonic - ns[GraphQL::Metrics::QUERY_START_TIME_MONOTONIC]
```
worth a comment here I think that we start the clock even before the instrumentation's `before_query`, this keeps catching me

eapache commented on lib/graphql/metrics/tracer.rb:
```diff
@@ -0,0 +1,110 @@
+        # chronological order.
+        case key
+        when GRAPHQL_GEM_LEXING_KEY
+          return setup_tracing_before_lexing { yield }
```
I thought we had a conversation about just passing the block directly instead of `{ yield }`, did this actually end up being the faster approach?

chrisbutcher commented:
> There's a lot of code here. At what point is it better to start splitting this up and shipping it to master in pieces? We don't have to release the 3.0 gem until all the pieces have merged.

Yeah that's a good point. The functional code: the comment-laden analyzer, tracer, and instrumentation weigh in at a mere 286 lines.

The rest is accounted for mainly by the README and integration_test.rb. The diff is also large because so much got deleted.

That said, I could technically split this up by shipping just the static analyzer on its own, stripping out all query/field timings... but at this stage in the game I think this being a total rewrite - which I paired with Scott on to do line-by-line review - it's probably best to just ship it as is, if there's no strong disagreement.

chrisbutcher commented:
cc: @swalkinshaw for another review at your convenience

swalkinshaw commented on graphql_metrics.gemspec:
```diff
@@ -1,11 +1,11 @@
 
 lib = File.expand_path("../lib", __FILE__)
 $LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
```
can we kill these too please?

Just use `require_relative` below

swalkinshaw commented on lib/graphql/metrics/analyzer.rb:
```diff
@@ -0,0 +1,125 @@
+        return if visitor.field_definition.introspection?
+        return if query.context[SKIP_FIELD_AND_ARGUMENT_METRICS]
+
+        # NOTE: @rmosolgo "I think it could be reduced to `arguments = visitor.arguments_for(ast_node)`"
```
Do we need this note anymore?

chrisbutcher replied:
Yeah, I'm going to write up some GH issues for some tech debt stuff discovered along the way. I think this comment will be useful to leave here until then.

swalkinshaw commented on lib/graphql/metrics/instrumentation.rb:
```diff
@@ -0,0 +1,51 @@
+    class Instrumentation
+      def before_query(query)
+        unless query_present_and_valid?(query)
+          query.context[GraphQL::Metrics::SKIP_GRAPHQL_METRICS_ANALYSIS] = true
```
Maybe just inline the two guard clauses and `return` early? I found it slightly confusing that you set the skip flag here instead of just directly returning

swalkinshaw commented on lib/graphql/metrics/tracer.rb:
```diff
@@ -0,0 +1,110 @@
+module GraphQL
+  module Metrics
+    class Tracer
+      # NOTE: Used to store timings from lexing, parsing, validation, before we have a context to store
```
Was there something under this note before? It seems out of place now
